chips/psoc62xa/src/scb.rs

```diff
@@ -0,0 +1,863 @@
+                    self.registers
+                        .tx_fifo_wr
+                        .write(TX_FIFO_WR::DATA.val(*buffer.get(0).unwrap() as u32));
+                    self.tx_buffer.put(Some(buffer));
+                    self.tx_length.set(tx_length);
+                    self.tx_position.set(0);
+                    self.enable_tx_interrupts();
+                    Ok(())
+                }
+                None => Err((ErrorCode::SIZE, buffer)),
+            }
+        }
+    }
+
+    pub fn receive_uart_async(
+        &self,
+        buffer: &'static mut [u8],
+        buffer_len: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        if self.rx_length.is_some() {
+            Err((ErrorCode::BUSY, buffer))
+        } else if buffer.len() < buffer_len || buffer_len == 0 {
+            Err((ErrorCode::SIZE, buffer))
+        } else {
+            match NonZeroUsize::new(buffer_len) {
+                Some(rx_length) => {
+                    self.enable_rx_interrupts();
+                    self.rx_buffer.put(Some(buffer));
+                    self.rx_length.set(rx_length);
+                    self.rx_position.set(0);
+                    Ok(())
+                }
+                None => Err((ErrorCode::SIZE, buffer)),
+            }
+        }
+    }
+}
+
+impl<'a> Transmit<'a> for Scb<'a> {
+    fn set_transmit_client(&self, client: &'a dyn TransmitClient) {
+        self.tx_client.set(client);
+    }
+
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+    ) -> Result<(), (kernel::ErrorCode, &'static mut [u8])> {
+        self.transmit_uart_async(tx_buffer, tx_len)
+    }
+
+    fn transmit_word(&self, _word: u32) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+
+    fn transmit_abort(&self) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+}
+
+impl<'a> Receive<'a> for Scb<'a> {
+    fn set_receive_client(&self, client: &'a dyn ReceiveClient) {
+        self.rx_client.set(client);
+    }
+
+    fn receive_buffer(
+        &self,
+        rx_buffer: &'static mut [u8],
+        rx_len: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        self.receive_uart_async(rx_buffer, rx_len)
+    }
+
+    fn receive_word(&self) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+
+    fn receive_abort(&self) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+}
+
+impl Configure for Scb<'_> {
+    fn configure(&self, params: kernel::hil::uart::Parameters) -> Result<(), ErrorCode> {
+        if params.baud_rate != 115200 || params.hw_flow_control {
+            Err(ErrorCode::NOSUPPORT)
+        } else {
+            // Modification of the SCB parameters require it to be disabled.
+            let enable = if self.registers.ctrl.is_set(CTRL::ENABLED) {
+                self.disable_scb();
+                true
+            } else {
+                false
+            };
```

Comment:
I think this is fine for now, but we should extend the UART HIL to be able to report `EBUSY` in this case. Silently interrupting an ongoing transmission for reconfiguration is not great.

Reply:
Opened #4304 for this.